### PR TITLE
Lily/mlm pretraining

### PR DESCRIPTION
**Pull request by lhz1029**

### CODE REVIEW

File: finetune/base.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ def _get_estimator_config(self):
             tf_random_seed=self.config.seed,
             save_summary_steps=self.config.val_interval,
             save_checkpoints_secs=None,
-            save_checkpoints_steps=None,
+            save_checkpoints_steps=10,
```
Comment (madisonmay) [inline finetune/base.py]: Let's keep the config the same and specify this in the document_representation project?

File: finetune/base_models/bert/featurizer.py [hunk 2]
```diff
@@ -119,3 +121,24 @@ def bert_featurizer(
...
+    **kwargs
+):
+    featurizer_state = bert_featurizer(X, encoder, config, train=train, reuse=reuse, **kwargs)
+    with tf.variable_scope("model/featurizer"):
+        embed_position(context, featurizer_state, config, train)
+        add_context_embed(featurizer_state)
+        hidden = featurizer_state['sequence_features']
+        w0, w = smooth_pos_attn(hidden, config, featurizer_state['lengths'])
+        text_embed = hidden[:, :, :config.n_embed]
+        seq_feats = tf.matmul(w, text_embed)
```
Comment (madisonmay) [inline finetune/base_models/bert/featurizer.py]: I think we should use a more flexible / general arch. for our pretraining experiments than we used for our finetuning experiments. IMO a reasonable baseline would be:

- Concat position embeddings to contextual embeddings output by base roberta
- Add 1-2 additional transformer layers that operate on the concatted vector
- Maybe worry about how we handle initialization (Can we make sure the new layers are dominated by residuals at the start so our hidden --> n_vocab projection for the MLM loss is still valid)?

File: finetune/util/in_memory_finetune.py [hunk 3]
```diff
@@ -35,8 +37,10 @@ def after_create_session(self, session, coord):
     def _evaluate(self, session):
         try:
             with tf.Graph().as_default():
-                from finetune import Classifier
-                model = Classifier(**self._config_to_finetune)
+                # from finetune import Classifier
```
Comment (madisonmay) [inline finetune/util/in_memory_finetune.py:40]: Little sketchy but I get what's going on.  Before merging to master we should fix this but fine for now.

File: finetune/base.py (hunk 1 above)
Comment (benleetownsend) [reply finetune/base.py]: Why do we need this change at all? This reactivates the tensorflow default savers which we don't normally use.

File: finetune/base_models/bert/featurizer.py [hunk 4]
```diff
@@ -119,3 +121,24 @@ def bert_featurizer(
...
+        featurizer_state['sequence_features'] = seq_feats
```
Comment (benleetownsend) [inline finetune/base_models/bert/featurizer.py]: This doesn't seem to give us correct classification features, can we either explicitly remove them or set them correctly

File: finetune/base_models/bert/model.py [hunk 5]
```diff
@@ -170,6 +170,23 @@ def get_encoder(cls, config=None, **kwargs):
             return RoBERTaEncoder(**kwargs)
 
 
+class RoBERTaPos(RoBERTa):
+    positional_info = True
+    # TODO: ideally we would pull out the positional logic so that
+    # it can be easily incorporated throughout all the SourceModels
+    featurizer = bert_featurizer_with_pos
+
+    @classmethod
```
Comment (benleetownsend) [inline finetune/base_models/bert/model.py]: Is there a reason these can not be inherited from RoBERTa, I don't see a difference?

File: finetune/model.py [hunk 6]
```diff
@@ -141,16 +141,26 @@ def _model_fn(features, labels, mode, params):
...
-                X,
-                encoder=encoder,
-                config=params,
-                train=train,
-                explain=build_explain,
-            )
-            if context is not None:
-                print("Embedding position")
-                embed_position(context, featurizer_state, params, train)
+            if params.base_model.positional_info:
```
Comment (benleetownsend) [inline finetune/model.py]: I think all the featurizers should have **kwargs, so they should be able to take extra keyword arguments they don't use.

Comment (benleetownsend) [reply finetune/model.py]: Or, as the extra layers just work on the featurizer state anyway we can just put them in here and therefore be able to apply them to any basemodel?

File: finetune/util/in_memory_finetune.py [hunk 7]
```diff
@@ -48,10 +52,10 @@ def _evaluate(self, session):
                     k: v for k, v in self._current_finetune.saver.fallback.items() if "global_step" not in k
                 }
                 train_x, train_y = self.train_data
-                model.fit(train_x, train_y)
+                model.fit(train_x, train_y, context=self.train_context)
                 test_x, test_y = self.test_data
-                test_accuracy = np.mean(model.predict(test_x) == test_y)
-                train_accuracy = np.mean(model.predict(train_x) == train_y)
+                test_accuracy = np.mean(model.predict(test_x, context=self.test_context) == test_y)
```
Comment (benleetownsend) [inline finetune/util/in_memory_finetune.py:58]: Pretty sure this metric isn't going to work

File: finetune/util/in_memory_finetune.py (hunk 3 above)
Comment (benleetownsend) [reply finetune/util/in_memory_finetune.py:40]: Which bit is sketchy?

File: finetune/base.py (hunk 1 above)
Comment (lhz1029) [reply finetune/base.py]: will change, especially know that I know about the cache_weights_to_file flag

File: finetune/base_models/bert/featurizer.py (hunk 4 above)
Comment (lhz1029) [reply finetune/base_models/bert/featurizer.py]: yeah whoops I meant to add a warning but forgot

File: finetune/model.py (hunk 6 above)
Comment (lhz1029) [reply finetune/model.py]: ah I like that!

File: finetune/util/in_memory_finetune.py (hunk 7 above)
Comment (lhz1029) [reply finetune/util/in_memory_finetune.py:58]: yeah I was seeing 0s across the board. what's wrong with this metric?

File: finetune/util/in_memory_finetune.py (hunk 3 above)
Comment (lhz1029) [reply finetune/util/in_memory_finetune.py:40]: yep agreed, I'll add a todo

File: finetune/util/in_memory_finetune.py (hunk 7 above)
Comment (benleetownsend) [reply finetune/util/in_memory_finetune.py:58]: Its testing exact match of the output dictionaries
